### PR TITLE
Use block device for container rootfs in case of devicemapper

### PR DESCRIPTION
**Pull request by amshinde**
In case of device mapper storage driver, the overlay file system is backed by a block device.
With this PR, we directly use the block device for container rootfs, instead of the overlay. As we do not go through 9pfs, this gives us much better I/O performance. We go through 9pfs only for the bind-mounts.

### CODE REVIEW

File: utils.go [hunk 1]
```diff
@@ -41,3 +43,37 @@ func fileCopy(srcPath, dstPath string) error {
...
+	if err != nil {
+		return nil, err
+	}
+
+	return b, nil
+}
+
+func reverseString(s string) string {
+	r := []rune(s)
+	for i, j := 0, len(r)-1; i < len(r)/2; i, j = i+1, j-1 {
```
Comment (jodh-intel) [inline utils.go]: Why not only call `len(r)` once?

File: container.go [hunk 2]
```diff
@@ -45,6 +45,21 @@ type ContainerStatus struct {
...
 
+// Mount describes a container mount.
+type Mount struct {
+	Source      string
+	Destination string
+
+	// Type specifies the type of filesystem to mount.
+	Type string
+
+	// Options list all the mount options of the filesystem.
```
Comment (jodh-intel) [inline container.go]: What is the order? Sorted or original ordering? Are duplicate options handled?

**Comment by jodh-intel** [comment]
@amshinde - I appreciate you have added some tests, but this is a big PR and it's not ideal that the test coverage is falling. Could you find a way to add some more to keep the coverage change `>= 0`? :-)

File: mount.go [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"syscall"
+)
+
+// These mounts need to be created by the agent within the VM
+var systemMounts = []string{"/proc", "/dev", "/dev/pts", "/dev/shm", "/dev/mqueue", "/sys", "/sys/fs/cgroup"}
```
Comment (sameo) [inline mount.go]: The order does not really matter, but I agree it could be better indented for easier maintenance.

**Comment by sameo** [comment]
Could you please rebase ? It will make the PR look smaller at least :)

**Comment by amshinde** [comment]
Rebased this PR.

**Comment by amshinde** [comment]
Added documentation pointing out the changes introduced with devicemapper and limitations.

File: pod.go [hunk 4]
```diff
@@ -998,6 +998,10 @@ func (p *Pod) checkContainersState(state stateString) error {
 	return nil
 }
 
+// Drive names are dependent on the order in which they are added to the hypervisor options.
+// driveIndex is an incrementing counter used to track the index at which current drive is added to the hypervisor.
+var driveIndex = int(0)
```
Comment (sboeuf) [inline pod.go]: We can do better than this, by adding this index to the hypervisor structure maybe. There is clearly no need for a global variable here. And more than that, keep in mind that we don't need to store this data because this will only be used by CreatePod(). Maybe this will be interesting to store in case we consider hotplugging some new block devices when the VM is already started.

File: container.go [hunk 5]
```diff
@@ -152,6 +152,34 @@ func (c *Container) SetPid(pid int) error {
...
+	if err != nil {
+		return err
+	}
+
+	c.state = tState
+
+	return nil
+}
+
+func (c *Container) setStateFstype(fstype string) error {
```
Comment (sboeuf) [inline container.go]: Ditto for this function.

File: README.md [hunk 6]
```diff
@@ -183,3 +183,17 @@ __Runtime network setup with CNI__
 4. Create bridge, TAP, and link all together with network interface previously created ([code](https://github.com/containers/virtcontainers/blob/0.5.0/network.go#L123-L205))
 
 5. Start VM inside the netns and start the container ([code](https://github.com/containers/virtcontainers/blob/0.5.0/api.go#L66-L70))
+
+## Storage
```
Comment (sboeuf) [inline README.md]: Thanks for updating the doc here. Could you also add the limitation that devicemapper block device can only be used when creating a new pod. Adding a new container (with a devicemapper rootfs) after we started the VM won't work.

File: mount.go [hunk 7]
```diff
@@ -122,3 +125,56 @@ func getDeviceForPath(path string) (device, error) {
...
+
+	file, err = os.Open(procMountsFile)
+	if err != nil {
+		return
+	}
+
+	defer file.Close()
+
+	reader := bufio.NewReader(file)
+	for {
```
Comment (amshinde) [inline mount.go]: @sboeuf /proc/mounts output can change, but the mount info for this device will not change(It has been mounted by docker and will be unmounted by it as well) Using /proc/mounts is a standard way to get mount info.

File: container.go [hunk 8]
```diff
@@ -152,6 +152,34 @@ func (c *Container) SetPid(pid int) error {
 	return c.storeProcess()
 }
 
+func (c *Container) setStateBlockIndex(index int) error {
+	tState := c.state
+	tState.BlockIndex = index
```
Comment (amshinde) [inline container.go]: I am just updating a temporary object here first. That way if the write to filesystem fails, the object in memory and filesystem are in sync

**Comment by amshinde** [comment]
@sboeuf I have pushed all your changes except for the call to addDrives(). Will wait for @sameo to comment on it.

File: README.md [hunk 9]
```diff
@@ -183,3 +183,19 @@ __Runtime network setup with CNI__
...
+
+Container workloads are shared with the virtualized environment through 9pfs.
+The devicemapper storage driver is a special case. The driver uses dedicated block devices rather than formatted filesystems, and operates at the block level rather than file level. We use this knowledge to directly use the underlying block device instead of the overlay file system for the container root file system. The block device maps to the top read-write layer for the overlay. This approach gives much better I/O performance as compared to using 9pfs to share the container file system.
```
Comment (jodh-intel) [inline README.md]: s/rather than file level/rather than the file level/

s/performance as compared/performance compared/

Also, this change mixes familiar and passive voices. I'd personally reword to allow you to drop all the "We"'s.

File: README.md [hunk 10]
```diff
@@ -183,3 +183,19 @@ __Runtime network setup with CNI__
...
+
+The approach above does introduce a limitation in terms of dynamic file copy in/out of the container via docker cp operations:
```
Comment (jodh-intel) [inline README.md]: formatting: `docker cp`.

What limitation? I think it would help to explain it first, then give examples.

File: README.md [hunk 11]
```diff
@@ -183,3 +183,19 @@ __Runtime network setup with CNI__
...
+
+```
+docker cp [OPTIONS] CONTAINER:SRC_PATH DEST_PATH|-
+docker cp [OPTIONS] SRC_PATH|- CONTAINER:DEST_PATH
```
Comment (jodh-intel) [inline README.md]: What are the pipe characters for?

File: README.md [hunk 12]
```diff
@@ -183,3 +183,19 @@ __Runtime network setup with CNI__
...
+```
+
+While the copy to container from host is not expected to work, it is possible to copy files from the container to host. We do need to sync the files to the block device before the copy operation.
```
Comment (jodh-intel) [inline README.md]: "from host ... to container" sounds more natural to me.

Overall, I find these two sentences confusing as it is unclear to the reader if they need to do anything "special": who needs to sync the files? virtcontainers, or is it up to the user to call `sync(1)` potentially multiple times to ensure the data is persisted?

File: README.md [hunk 13]
```diff
@@ -183,3 +183,19 @@ __Runtime network setup with CNI__
...
+
+The devicemapper block device can only be used when creating a pod. Adding a container with a devicemapper rootfs after the VM has started does not work currently.
```
Comment (jodh-intel) [inline README.md]: - What behaviour will the user see if they try to add a devicemapper rootfs after pod creation? Is it handled elegantly, or do you get a crash / odd error message?
- Will it work in the future? (Is there an issue for this?)
- How can you tell if a pod was created with a devicemapper rootfs after the event?

File: hyperstart.go [hunk 14]
```diff
@@ -323,7 +323,9 @@ func (h *hyper) bindUnmountContainerRootfs(podID, cID string) error {
 func (h *hyper) bindUnmountAllRootfs(pod Pod) {
 	for _, c := range pod.containers {
 		h.bindUnmountContainerMounts(c.mounts)
-		h.bindUnmountContainerRootfs(pod.id, c.id)
+		if c.state.Fstype == "" {
+			h.bindUnmountContainerRootfs(pod.id, c.id)
```
Comment (jodh-intel) [inline hyperstart.go]: Missing error check on this call.

File: hyperstart.go [hunk 15]
```diff
@@ -342,6 +344,23 @@ func (h *hyper) init(pod *Pod, config interface{}) (err error) {
 	// Override pod agent configuration
 	pod.config.AgentConfig = h.config
 
+	created, err := pod.isPodCreated()
+	if err != nil {
+		return err
+	}
+
+	if !created {
+		virtLog.Infof("Pod %s has not competed creation yet, adding hypervisor devices", pod.id)
```
Comment (jodh-intel) [inline hyperstart.go]: typo: "completed". I'd write as, `Pod %s not fully created yet, adding hypervisor devices`

File: mount.go [hunk 16]
```diff
@@ -32,3 +41,189 @@ func isSystemMount(m string) bool {
...
+		var line string
+
+		line, err = reader.ReadString('\n')
+		if err == io.EOF {
+			err = fmt.Errorf("Mount %s not found", mountPoint)
+			break
```
Comment (jodh-intel) [inline mount.go]: This still looks a little odd to me - I'd just `return fmt.Errorf(...)` for all these cases as a return cleary breaks out of the loop.

File: mount.go [hunk 17]
```diff
@@ -32,3 +41,189 @@ func isSystemMount(m string) bool {
...
+		}
+
+		if mountPoint == fields[procPathIndex] {
+			devicePath = fields[procDeviceIndex]
+			fsType = fields[procTypeIndex]
+			return
+		}
+	}
+
+	return
```
Comment (jodh-intel) [inline mount.go]: If you immediately return an error at the call site, this would presumably need to become a catch-all error.

File: mount.go [hunk 18]
```diff
@@ -32,3 +41,189 @@ func isSystemMount(m string) bool {
...
+
+// getVirtBlockDriveName returns the disk name format for virtio-blk
+// Reference: https://github.com/torvalds/linux/blob/master/drivers/block/virtio_blk.c @c0aa3e0916d7e531e69b02e426f7162dfb1c6c0
+func getVirtDriveName(index int) (string, error) {
+	if index < 0 {
+		return "", fmt.Errorf("Index cannot be negative for drive")
+	}
+
+	// Prefix used for virtio-block devices
+	prefix := "vd"
```
Comment (jodh-intel) [inline mount.go]: This could be `const`

File: mount_test.go [hunk 19]
```diff
@@ -41,3 +49,132 @@ func TestIsSystemMount(t *testing.T) {
...
+
+	expected := "/proc"
+
+	if dev.mountPoint != expected {
+		t.Fatalf("Expected %s mountpoint, got %s", expected, dev.mountPoint)
+	}
+}
+
+func TestGetDeviceForPathEmptyPath(t *testing.T) {
+	_, err := getDeviceForPath("")
```
Comment (jodh-intel) [inline mount_test.go]: I think there is scope to add some more tests for this function:

- `getDeviceForPath("///")`
- `getDeviceForPath("/../../.././././../.")`
- what if the specified path has spaces / special characters / unicode?
- create a test that creates a bind mount manually and see if `getDevicePath()` returns the expected value.

File: mount.go [hunk 20]
```diff
@@ -32,3 +41,189 @@ func isSystemMount(m string) bool {
...
+			fsType = fields[procTypeIndex]
+			return
+		}
+	}
+
+	return
+}
+
+// isDeviceMapper checks if the device with the major and minor numbers is a devicemapper block device
+func isDeviceMapper(major, minor int) (bool, error) {
```
Comment (jodh-intel) [inline mount.go]: Missing tests. You could create a variable this function uses (something like, `var formatTemplate  := /sys/dev/block/%d:%d/dm`), and have the tests modify that so you can check the expected return value.

File: mount_test.go [hunk 21]
```diff
@@ -41,3 +49,132 @@ func TestIsSystemMount(t *testing.T) {
...
+		t.Fatal(err)
+	}
+
+	if path != "proc" || fstype != "proc" {
+		t.Fatal(err)
+	}
+}
+
+func TestGetVirtDriveNameInvalidIndex(t *testing.T) {
+	index := -1
```
Comment (jodh-intel) [inline mount_test.go]: You could drop this var and just pass `-1` directly as it's only used once.

File: hyperstart.go (hunk 14 above)
Comment (amshinde) [inline hyperstart.go]: I have just wrapped the existing call in an if condition. This will require changes up the stack, something that I want to do in a separate PR.

**Comment by amshinde** [comment]
Rebased this PR, addressed review comments, fixed documentation and added some more tests.
@jodh-intel , @sboeuf , @sameo Please have a look.

File: README.md [hunk 22]
```diff
@@ -183,3 +183,23 @@ __Runtime network setup with CNI__
...
+
+## Storage
+
+Container workloads are shared with the virtualized environment through 9pfs.
+The devicemapper storage driver is a special case. The driver uses dedicated block devices rather than formatted filesystems, and operates at the block level rather than the file level. This knowledge has been used to directly use the underlying block device instead of the overlay file system for the container root file system. The block device maps to the top read-write layer for the overlay. This approach gives much better I/O performance compared to using 9pfs to share the container file system.
+
+The approach above does introduce a limitation in terms of dynamic file copy in/out of the container via `docker cp` operations.
+The copy operation from host to container accesses the mounted file system on the host side. This is not expected to work and may lead to inconsistencies as the block device will be simultaneously written to, from two different mounts.
+The copy operation from container to host will work, provided the user calls `sync(1)` prior to the copy to make sure any outstanding cached data is written to the block device.
```
Comment (jodh-intel) [inline README.md]: Where should the `sync` be called? In the VM, or on the host? Should the container be paused, then the `sync` call made I wonder?

File: hyperstart.go (hunk 14 above)
Comment (jodh-intel) [inline hyperstart.go]: Is there an issue to track this? Not checking a return value for an operation like this sounds dangerous to me. I'd add a reference to an issue in a comment for this line.

File: README.md (hunk 22 above)
Comment (amshinde) [inline README.md]: @jodh-intel I have added "from within the container" to make this clearer. And no, we do not need to pause the container.

**Comment by jodh-intel** [comment]
lgtm

**Comment by amshinde** [comment]
@jodh-intel I have created a separate issue for checking the return value of unmount: https://github.com/containers/virtcontainers/issues/295.
Please take a look.

**Comment by jodh-intel** [comment]
Thanks @amshinde.

File: mount_test.go [hunk 23]
```diff
@@ -41,3 +51,227 @@ func TestIsSystemMount(t *testing.T) {
...
+
+		data := bytes.Split(output, []byte(","))
+		if len(data) < 2 {
+			t.Fatal()
+		}
+
+		majorStr := strings.TrimSpace(string(data[0]))
+		minorStr := strings.TrimSpace(string(data[1]))
+
+		majorNo, err := strconv.Atoi(majorStr)
```
Comment (sboeuf) [inline mount_test.go]: mount_test.go:75:12: ineffectual assignment to err

You should do something with err, or just use `_`

File: mount_test.go [hunk 24]
```diff
@@ -41,3 +51,227 @@ func TestIsSystemMount(t *testing.T) {
...
+		minorNo, err := strconv.Atoi(minorStr)
```
Comment (sboeuf) [inline mount_test.go]: mount_test.go:76:12: ineffectual assignment to err

You should do something with err, or just use `_`

File: container.go [hunk 25]
```diff
@@ -238,13 +260,10 @@ func (c *Container) setContainerState(state stateString) error {
 		return errNeedState
 	}
 
-	// update in-memory state
```
Comment (amshinde) [inline container.go]: fixed

File: mount_test.go (hunk 23 above)
Comment (amshinde) [inline mount_test.go]: fixed

File: mount_test.go (hunk 24 above)
Comment (amshinde) [inline mount_test.go]: fixed

**Comment by sboeuf** [comment]
LGTM
